File: C++/Sorting/IntroSort.cpp
//
//  IntroSort.cpp
//  Sorting
//
//  Created by Paul Kraft on 10.04.16.
//  Copyright © 2016 pauljohanneskraft. All rights reserved.
//

#include "IntroSort.hpp"
#include "InsertionSort.hpp"
#include "ShellSort.hpp"
#include "BinaryTreeSort.hpp"
#include <cmath>

IntroSort::IntroSort() : QuickSort("IntroSort\t\t\t") {}

void IntroSort::sort(Comparable *array[], int length) {
    sort(array, 0, length - 1);
}

void IntroSort::sort(Comparable *array[], int left, int right) {
    sort(array, left, right, log(right-left));
}

void IntroSort::sort(Comparable *array[], int left, int right, int depth) {
    if(right - left < 20) {
        InsertionSort ins = *new InsertionSort();
        ins.sort(array, left, right);
        return;
    }
    if(depth == 0) {
        Sort *s = new BinaryTreeSort();
        s->sort(array, left, right);
        return;
    }
    
    int pivot = partition(array, left, right);
    sort(array, left, pivot - 1);
    sort(array, pivot, right);
}

File: Objective-C/Sorting/sorting algorithms/IntroSort.m
//
//  IntroSort.m
//  Sorting
//
//  Created by Paul Kraft on 14.04.16.
//  Copyright © 2016 pauljohanneskraft. All rights reserved.
//

#import "IntroSort.h"
#import "SelectionSort.h"
#import "HeapSort.h"
#import "BinaryTreeSort.h"

@implementation IntroSort

+ (void) sort: (NSMutableArray *)array fromLeft: (long)start toRight: (long)end {
    [self sort:array fromLeft:start toRight:end withDepth:log(array.count)];
}

+ (void) sort: (NSMutableArray *)array fromLeft: (long)start toRight: (long)end withDepth:(long)depth {
    if( start >= end ) return;
    if( end - start < 20 ) {
        [SelectionSort sort:array fromLeft:start toRight:end];
        return;
    }
    if( depth < 0 ) {
        [BinaryTreeSort sort:array fromLeft:start toRight:end];
        //[HeapSort sort:array fromLeft:start toRight:end];
        return;
    }
    long pivot = [QuickSort partition:array fromLeft:start toRight:end];
    [self sort:array fromLeft:start toRight:pivot - 1 withDepth: depth - 1];
    [self sort:array fromLeft:pivot toRight:end withDepth:depth - 1];
}

@end
